File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
N, K = list(map(int, input().split()))
weight = [0] * N
price = [0] * N
dp = [[0] * (K + 1) for _ in range(N)]

for i in range(N):
    w, p = list(map(int, input().split()))
    weight[i] = w
    price[i] = p


for i in range(N):
    for j in range(K + 1):
        if weight[i] > j:  # 현재 가방의 무게보다 물건의 무게가 크면
            dp[i][j] = dp[i - 1][j]  # 넣지 못하므로 이전 가치를 그대로 쓴다
        else:  # 현재 가방의 무게와 물건의 무게가 같거나 더 작다면
            dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - weight[i]] + price[i])


print(dp[N - 1][K])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, K] = input[0].split(" ").map(Number);

const items = [];
const dp = Array(K + 1).fill(0);

for (let i = 1; i <= N; i++) {
  const [W, V] = input[i].split(" ").map(Number);

  items.push([W, V]);
}

for (let i = 1; i <= N; i++) {
  const [W, V] = items[i - 1];

  for (let j = K; j >= W; j--) {
    dp[j] = Math.max(dp[j], dp[j - W] + V);
  }
}

console.log(dp[K]);
